packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts

```diff
@@ -951,50 +1157,110 @@ export class AutoScalingGroup extends AutoScalingGroupBase implements
       throw new Error('Setting \'initOptions\' requires that \'init\' is also set');
     }
 
-    this.securityGroup = props.securityGroup || new ec2.SecurityGroup(this, 'InstanceSecurityGroup', {
-      vpc: props.vpc,
-      allowAllOutbound: props.allowAllOutbound !== false,
-    });
-    this.connections = new ec2.Connections({ securityGroups: [this.securityGroup] });
-    this.securityGroups.push(this.securityGroup);
-    Tags.of(this).add(NAME_TAG, this.node.path);
-
-    this.role = props.role || new iam.Role(this, 'InstanceRole', {
-      roleName: PhysicalName.GENERATE_IF_NEEDED,
-      assumedBy: new iam.ServicePrincipal('ec2.amazonaws.com'),
-    });
-
-    this.grantPrincipal = this.role;
-
     if (props.groupMetrics) {
       this.groupMetrics.push(...props.groupMetrics);
     }
 
-    const iamProfile = new iam.CfnInstanceProfile(this, 'InstanceProfile', {
-      roles: [this.role.roleName],
-    });
+    let launchConfig: CfnLaunchConfiguration | undefined = undefined;
+    if (props.launchTemplate || props.mixedInstancesPolicy) {
+      if (props.machineImage) {
+        throw new Error('Setting \'machineImage\' must not be set when \'launchTemplate\' or \'mixedInstancesPolicy\' is set');
+      }
+      if (props.instanceType) {
+        throw new Error('Setting \'instanceType\' must not be set when \'launchTemplate\' or \'mixedInstancesPolicy\' is set');
+      }
+      if (props.role) {
+        throw new Error('Setting \'role\' must not be set when \'launchTemplate\' or \'mixedInstancesPolicy\' is set');
+      }
+      if (props.userData) {
+        throw new Error('Setting \'userData\' must not be set when \'launchTemplate\' or \'mixedInstancesPolicy\' is set');
+      }
+      if (props.securityGroup) {
+        throw new Error('Setting \'securityGroup\' must not be set when \'launchTemplate\' or \'mixedInstancesPolicy\' is set');
+      }
+      if (props.keyName) {
+        throw new Error('Setting \'keyName\' must not be set when \'launchTemplate\' or \'mixedInstancesPolicy\' is set');
+      }
+      if (props.instanceMonitoring) {
+        throw new Error('Setting \'instanceMonitoring\' must not be set when \'launchTemplate\' or \'mixedInstancesPolicy\' is set');
+      }
+      if (props.associatePublicIpAddress) {
+        throw new Error('Setting \'associatePublicIpAddress\' must not be set when \'launchTemplate\' or \'mixedInstancesPolicy\' is set');
+      }
+      if (props.spotPrice) {
+        throw new Error('Setting \'spotPrice\' must not be set when \'launchTemplate\' or \'mixedInstancesPolicy\' is set');
+      }
+      if (props.blockDevices) {
+        throw new Error('Setting \'blockDevices\' must not be set when \'launchTemplate\' or \'mixedInstancesPolicy\' is set');
+      }
```

Comment:
I won't move the validations in the other branch for launch config btw, because they are short, and some subsequent code requires the non-null check there.